portal-impl/src/com/liferay/portal/security/auth/AuthVerifierPipeline.java

```diff
@@ -64,23 +65,53 @@ public static String getAuthVerifierPropertyName(String className) {
 			PropsKeys.AUTH_VERIFIER, simpleClassName, StringPool.PERIOD);
 	}
 
-	public static AuthVerifierResult verifyRequest(
+	public AuthVerifierPipeline(Map<String, Object> filterProperties) {
+		_filterProperties = filterProperties;
+
+		AuthVerifierTrackerCustomizer.addAuthVerifierPipeline(this);
+	}
+
+	public AuthVerifierResult verifyRequest(
 			AccessControlContext accessControlContext)
 		throws PortalException {
 
-		return _authVerifierPipeline._verifyRequest(accessControlContext);
-	}
+		if (accessControlContext == null) {
+			throw new IllegalArgumentException(
+				"Access control context is null");
+		}
 
-	private AuthVerifierPipeline() {
-		Registry registry = RegistryUtil.getRegistry();
+		HttpServletRequest httpServletRequest =
+			accessControlContext.getRequest();
 
-		Filter filter = registry.getFilter(
-			"(objectClass=" + AuthVerifier.class.getName() + ")");
+		String requestURI = httpServletRequest.getRequestURI();
 
-		_serviceTracker = registry.trackServices(
-			filter, new AuthVerifierTrackerCustomizer());
+		String contextPath = httpServletRequest.getContextPath();
 
-		_serviceTracker.open();
+		if (requestURI.equals(contextPath)) {
+			requestURI += "/";
+		}
+		else {
+			requestURI = requestURI.substring(contextPath.length());
+		}
+
+		Set<AuthVerifierConfiguration> excludeAuthVerifierConfigurations =
+			new HashSet<>();
+
+		_excludeURLPatternMapper.consumeValues(
+			excludeAuthVerifierConfigurations::addAll, requestURI);
+
+		AuthVerifierConfigurationConsumer authVerifierConfigurationConsumer =
+			new AuthVerifierConfigurationConsumer(
+				accessControlContext, excludeAuthVerifierConfigurations);
+
+		_includeURLPatternMapper.consumeValues(
+			authVerifierConfigurationConsumer, requestURI);
+
```

Comment:
hey @arthurchan35,

should we first test for `url.include` and only if we have a match then test for `url.exclude`? I believe we should only do it when we have a match, shouldn't we?

Reply:
https://github.com/liferay-appsec/liferay-portal/pull/186/commits/bb09712d8e671de3a58faaec43a82c2aa89b53db